src/member.ts

```diff
@@ -0,0 +1,358 @@
+import { randomBytes } from "crypto";
+import { EventEmitter } from "events";
+
+export type ResponseType = "accept" | "deny" | "cancel" | "conflict";
+export type ResponseState =
+  | "pending"
+  | "ready"
+  | "cancelled"
+  | "failed"
+  | "finished";
+export type Operation = "add" | "remove" | "merge";
+export type RequestIdentifier = "request";
+export type ResponseIdentifier = "response";
+export type FeedItem = Request | Response;
+
+export type ID = string;
+
+export const REQUEST_TYPE = "request" as RequestIdentifier;
+export const RESPONSE_TYPE = "response" as ResponseIdentifier;
+
+interface RawRequest {
+  // Used to differentiate between req/res
+  type: RequestIdentifier;
+  // ID of the request
+  id: ID;
+  // ID of the creator
+  from: ID;
+  // What sort of operation this is
+  operation: Operation;
+}
+
+export interface WhoRequest extends RawRequest {
+  // Who to add or remove
+  who: ID;
+}
+
+export interface MergeRequest extends RawRequest {
+  // Which requests have been merged for this one
+  toMerge: Request[];
+}
+
+export type Request = WhoRequest | MergeRequest;
+
+export interface Response {
+  // Used to differentiate between req/res
+  type: ResponseIdentifier;
+  // ID of the request
+  id: ID;
+  // ID of the request creator
+  from: ID;
+  // Our response to this request
+  response: ResponseType;
+}
+
+export class RequestState {
+  req: Request;
+  signatures: {
+    [id in ID]: Response;
+  };
+  finished: boolean;
+
+  constructor(req: Request) {
+    this.req = req;
+    this.signatures = {};
+    this.finished = false;
+  }
+
+  getState(
+    neededSignatures: number,
+    maxDenied: number = neededSignatures
+  ): ResponseState {
+    if (this.finished) return "finished";
+    if (this.isCancelled()) return "cancelled";
+    if (this.isConflicted()) return "failed";
+    if (this.numberAccepted() >= neededSignatures) return "ready";
+    if (this.numberDenied() >= maxDenied) return "failed";
+
+    return "pending";
+  }
+
+  addResponse(author: ID, response: Response) {
+    this.signatures[author] = response;
+  }
+
+  get id() {
+    return this.req.id;
+  }
+
+  get from() {
+    return this.req.from;
+  }
+
+  get operation() {
+    return this.req.operation;
+  }
+
+  numberDenied(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "deny";
+    }).length;
+  }
+
+  numberAccepted(): number {
+    return Object.keys(this.signatures).filter((id: ID) => {
+      return this.signatures[id].response === "accept";
+    }).length;
+  }
+
+  isConflicted(): boolean {
+    return Object.keys(this.signatures).some((id: ID) => {
+      return this.signatures[id].response === "conflict";
+    });
+  }
+
+  isCancelled(): boolean {
+    return this.signatures[this.from]?.response === ("cancel" as ResponseType);
```

Comment:
This should work without the casting.

Reply:
Thank you. Still getting used to what is and isn't limited by TS. 😅